executor/common_zlib.h

```diff
@@ -0,0 +1,540 @@
+// provision of more input data or more output space.  In the case of invalid
+// inflate data (a negative error), the dest and source pointers are updated to
+// facilitate the debugging of deflators.
+
+// The return codes are:
+
+//   2:  available inflate data did not terminate
+//   1:  output space exhausted before completing inflate
+//   0:  successful inflate
+//  -1:  invalid block type (type == 3)
+//  -2:  stored block length did not match one's complement
+//  -3:  dynamic block code description: too many length or distance codes
+//  -4:  dynamic block code description: code lengths codes incomplete
+//  -5:  dynamic block code description: repeat lengths with no first length
+//  -6:  dynamic block code description: repeat more than specified lengths
+//  -7:  dynamic block code description: invalid literal/length code lengths
+//  -8:  dynamic block code description: invalid distance code lengths
+//  -9:  dynamic block code description: missing end-of-block code
+// -10:  invalid literal/length or distance code in fixed or dynamic block
+// -11:  distance is too far back in fixed or dynamic block
+static int puff(
+    unsigned char* dest, // pointer to destination pointer
+    unsigned long* destlen, // amount of output space
+    const unsigned char* source, // pointer to source data pointer
+    unsigned long* sourcelen) // amount of input available
+{
+	struct puff_state s; // input/output state
+	int last, type; // block information
+	int err; // return value
+
+	// initialize output state
+	s.out = dest;
+	s.outlen = *destlen; // ignored if dest is null
+	s.outcnt = 0;
+
+	// initialize input state
+	s.in = source;
+	s.inlen = *sourcelen;
+	s.incnt = 0;
+	s.bitbuf = 0;
+	s.bitcnt = 0;
+
+	// return if bits() or decode() tries to read past available input
+	if (setjmp(s.env) != 0) // if came back here via longjmp()
+		err = 2; // then skip do-loop, return error
+	else {
+		// process blocks until last block or error
+		do {
+			last = puff_bits(&s, 1); // one if last block
+			type = puff_bits(&s, 2); // block type 0..3
+			err = type == 0 ? puff_stored(&s) : (type == 1 ? puff_fixed(&s) : (type == 2 ? puff_dynamic(&s) : -1)); // type == 3, invalid
+			if (err != 0)
+				break; // return with error
+		} while (!last);
+	}
+
+	// update the lengths and return
+	if (err <= 0) {
+		*destlen = s.outcnt;
+		*sourcelen = s.incnt;
+	}
+	return err;
+}
+
+//% END CODE DERIVED FROM puff.{c,h}
+
+#include <sys/mman.h>
+#define ZLIB_HEADER_WIDTH 2 // Two-byte zlib header width.
+
+static int puff_zlib_to_file(
+    const unsigned char* source,
+    unsigned long sourcelen,
+    int dest_fd,
+    unsigned long destlen)
+{
+	// Ignore zlib header.
+	if (sourcelen < ZLIB_HEADER_WIDTH)
+		return -12; // use next available negative return value
+	source += ZLIB_HEADER_WIDTH;
+	sourcelen -= ZLIB_HEADER_WIDTH;
+
+	// Memory-map destination file dest_fd.
+	void* ret = mmap(0, destlen, PROT_WRITE, MAP_SHARED, dest_fd, 0);
```

Comment:
I see that on line 291 the code is reading from the dest buffer:
```
					s->out[s->outcnt] =
#ifdef INFLATE_ALLOW_INVALID_DISTANCE_TOOFAR_ARRR
					    dist > s->outcnt ? 0 :
#endif
							     s->out[s->outcnt - dist];
```


Thus we need PROT_READ as well.